File: Assets/src/model/grasses/Tunnelroot.cs
using System;
using System.Linq;
using UnityEngine;

[System.Serializable]
[ObjectInfo(description: "Walking into the Tunnelroot teleports you to the paired Tunnelroot elsewhere on this level.")]
public class Tunnelroot : Grass, IActorEnterHandler {
  public static bool CanOccupy(Tile tile) => tile is Ground;

  private Tunnelroot partner;
  private bool isOpen = true;
  [field:NonSerialized] /// controller only
  public event Action<bool> OnOpenChanged;

  public Tunnelroot(Vector2Int pos) : base(pos) {
  }

  public bool IsOpen() => isOpen;

  public void PartnerWith(Tunnelroot other) {
    if (other.partner != null) {
      Debug.LogWarning("Partnering an already partnered Tunnelroot!");
    }
    other.partner = this;
    this.partner = other;
  }

  protected override void HandleLeaveFloor() {
    base.HandleLeaveFloor();
    if (partner != null && partner.floor != null) {
      // let my floor = null code finish first and *then* remove partner, otherwise we recurse infinitely
      GameModel.main.EnqueueEvent(() => {
        partner.floor.Remove(partner);
      });
    }
  }

  public void HandleActorEnter(Actor who) {
    if (partner == null) {
      Debug.LogWarning("Unpartnered Tunnelroot!");
    }

    if (!partner.IsDead && partner.body == null && isOpen) {
      // Close();
      // partner.Close();
      var newTile = floor.BreadthFirstSearch(partner.pos, (tile) => true).Skip(1).Where(t => t.CanBeOccupied()).FirstOrDefault();
      if (newTile != null) {
        OnNoteworthyAction();
        partner.OnNoteworthyAction();
        who.pos = newTile.pos;
        who.statuses.Add(new SurprisedStatus());
      }
    }
  }

  private void Close() {
    isOpen = false;
    OnOpenChanged?.Invoke(isOpen);
    AddTimedEvent(5, Reopen);
    OnNoteworthyAction();
  }

  void Reopen() {
    isOpen = true;
    OnOpenChanged?.Invoke(isOpen);
  }
}

File: web/src/model/grasses/Tunnelroot.ts
import { Grass } from './Grass';
import { ACTOR_ENTER_HANDLER, type IActorEnterHandler } from '../../core/types';
import { Ground } from '../Tile';
import { GameModelRef } from '../GameModelRef';
import { SurprisedStatus } from '../tasks/SleepTask';
import { entityRegistry } from '../../generator/entityRegistry';
import type { Vector2Int } from '../../core/Vector2Int';
import type { Tile } from '../Tile';

/**
 * Paired teleport grass. Walking into one teleports you near the partner.
 * Enemies get SurprisedStatus on teleport.
 * Port of C# Tunnelroot.cs.
 */
export class Tunnelroot extends Grass implements IActorEnterHandler {
  readonly [ACTOR_ENTER_HANDLER] = true as const;

  private partner: Tunnelroot | null = null;
  private isOpen = true;

  static canOccupy(tile: Tile): boolean {
    return tile instanceof Ground;
  }

  constructor(pos: Vector2Int) {
    super(pos);
  }

  partnerWith(other: Tunnelroot): void {
    other.partner = this;
    this.partner = other;
  }

  protected handleLeaveFloor(): void {
    if (this.partner && this.partner.floor) {
      const partner = this.partner;
      GameModelRef.main.enqueuEvent(() => {
        partner.floor?.remove(partner);
      });
    }
  }

  handleActorEnter(who: any): void {
    if (!this.partner || this.partner.isDead || !this.isOpen) return;

    // Check partner tile is unoccupied
    const partnerBody = this.floor!.bodies.get(this.partner.pos);
    if (partnerBody) return;

    // Find nearest open tile near partner via BFS (skip partner pos itself)
    let destTile: Tile | null = null;
    let first = true;
    for (const t of this.floor!.breadthFirstSearch(this.partner.pos)) {
      if (first) { first = false; continue; } // Skip(1) — skip partner pos
      if (t.canBeOccupied() && !(t.grass instanceof Tunnelroot)) {
        destTile = t;
        break;
      }
    }

    if (destTile) {
      this.onNoteworthyAction();
      this.partner.onNoteworthyAction();
      who.pos = destTile.pos;
      if (who !== GameModelRef.main.player) {
        who.statuses.add(new SurprisedStatus());
      }
    }
  }
}

entityRegistry.register('Tunnelroot', Tunnelroot);
